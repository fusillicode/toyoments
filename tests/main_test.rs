use std::process::Command;

#[test]
fn main_processes_transactions_without_errors_as_expected() {
    let bin = env!("CARGO_BIN_EXE_toyments");
    let csv_path = "tests/fixtures/main_processes_transactions_without_errors_as_expected.csv";

    let output = Command::new(bin).arg(csv_path).output().unwrap();
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    assert!(
        output.status.success(),
        "binary failed: status={:?} stderr={stderr} stdout={stdout}",
        output.status,
    );
    insta::assert_snapshot!(stdout);
}

#[test]
fn main_processes_transactions_with_errors_as_expected() {
    let bin = env!("CARGO_BIN_EXE_toyments");
    let csv_path = "tests/fixtures/main_processes_transactions_with_errors_as_expected.csv";

    let output = Command::new(bin).arg(csv_path).output().unwrap();
    let stdout = String::from_utf8_lossy(&output.stdout);
    let stderr = String::from_utf8_lossy(&output.stderr);

    assert_eq!(Some(1), output.status.code());
    insta::assert_snapshot!(stdout);
    // Not using snapshotting because I consider errors current representation not stable enough.
    // Core deserialization error for invalid type
    assert!(stderr.contains("error deserializing transaction"));
    assert!(stderr.contains("unknown variant `foo`"));
    assert!(stderr.contains("TransactionAlreadyDisputed"));
    assert!(stderr.contains("TransactionNotFound"));
    assert!(stderr.contains("TransactionNotDisputed"));
    assert!(stderr.contains("InsufficientFunds"));
    assert!(stderr.contains("ClientAccountLocked"));
}
