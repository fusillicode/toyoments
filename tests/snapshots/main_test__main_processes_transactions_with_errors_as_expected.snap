---
source: tests/main_test.rs
assertion_line: 37
expression: stdout
---
client_id,available,held,total,locked
1,4.0,0.0,4.0,false
2,1.0,0.0,1.0,true
